modules/@apostrophecms/settings/index.js

```diff
@@ -0,0 +1,207 @@
+module.exports = {
+  options: {
+    alias: 'settings',
+    subforms: {},
+    // XXX not implemented until Phase 2
+    groups: {}
+  },
+
+  init(self) {
+    self.userSchema = [];
+    self.subforms = [];
+    self.initSubforms();
+    self.enableBrowserData();
+    self.addToAdminBar();
+  },
+
+  handlers(self) {
+    return {
+      'apostrophe:modulesRegistered': {
+        addModal() {
+          self.addSettingsModal();
+        }
+      }
+    };
+  },
+
+  methods(self) {
+    return {
+      hasSchema() {
+        return self.userSchema.length > 0;
+      },
+
+      // Initialize the subforms configuration
+      initSubforms() {
+        self.userSchema = self.inferUserSchema();
+
+        for (const [ name, config ] of Object.entries(self.options.subforms)) {
+          const schema = self.getSubformSchema(name);
+          const label = config.label || schema[0]?.label;
+          self.subforms.push({
+            ...config,
+            name,
+            label,
+            schema
+          });
+        }
+      },
+
+      // Get the subset of the user schema that is relevant to the configured
+      // subforms.
+      inferUserSchema() {
+        const schema = [];
+        const subforms = self.options.subforms;
+        const userSchema = self.apos.user.schema;
+        const allSettingsFields = [
+          ...new Set(
+            Object.keys(subforms)
+              .reduce((acc, subform) => {
+                return acc.concat(subforms[subform].fields);
+              }, [])
+          )
+        ];
+        self.validateSettingsSchema(allSettingsFields, userSchema);
+
+        for (const field of userSchema) {
+          if (allSettingsFields.includes(field.name)) {
+            schema.push(field);
+          }
+        }
+        return schema;
+      },
+
+      // Validate that the fields configured in the settings module exist in the
+      // user schema and are not forbidden.
+      // XXX Temporary (Phase 2,3) extend the forbidden protected fields.
+      validateSettingsSchema(settingsFieldNames, userSchema) {
+        const forbiddenFields = [
+          'role',
+          // These will be allowed in Phase 2 and 3
+          'username',
+          'password',
+          'email'
+        ];
+        for (const name of settingsFieldNames) {
+          if (!userSchema.some(field => field.name === name)) {
+            throw new Error(`[@apostrophecms/settings] The field "${name}" is not a valid user field.`);
+          }
+          if (forbiddenFields.includes(name)) {
+            throw new Error(`[@apostrophecms/settings] The field "${name}" is forbidden.`);
+          }
+        }
+      },
+
+      // Get subform fields by subform name.
+      getSubformSchema(name) {
+        const subform = self.options.subforms[name];
+        if (!subform) {
```

Comment:
How about a typeof check as well, making sure it is an object and has a fields array property?

Reply:
Good catch! I'll do that now.

Reply:
No need to check here, instead we are doing it on a boot time (malformed `fields`). Additionally we throw on unknown field in `fields`. So it's a boot check and panic, getting it later is completely safe (should be).